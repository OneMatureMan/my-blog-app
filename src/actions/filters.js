export const filterByText = (text) => ({
    type: 'FILTER_BY_TEXT',
    text
});


export const sortByDate = () => ({
    type: 'SORT_BY_DATE'
});


export const sortByTitle = () => ({
    type: 'SORT_BY_TITLE'
});