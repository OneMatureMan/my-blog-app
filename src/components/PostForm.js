import React from 'react';


class PostForm extends React.Component {
    state = {
        title: this.props.post ? this.props.post.title : '',
        body: this.props.post ? this.props.post.body : '',
        error: ''
    }


    handleTitleChange = (e) => {
        const title = e.target.value;
        this.setState(({ title }))
    }
    
    handleBodyChange = (e) => {
        const body = e.target.value;
        this.setState(({ body }))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.title || !this.state.body) {
            return this.setState(({ error: 'Please enter both title and body'}))
        }
        this.setState(({ error: '' }))
        this.props.onSubmit({
            title: this.state.title,
            body: this.state.body
        })        
    }
    
    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input 
                    type="text"
                    name='title'
                    placeholder='Title'
                    autoFocus
                    value = {this.state.title}
                    onChange={this.handleTitleChange}
                />
                <textarea
                    name='body'
                    placeholder='Enter the blogs body here'
                    value= {this.state.body}
                    onChange={this.handleBodyChange}
                />
                <button>Save Post</button>
            </form>
        )
    }
}


export default PostForm;