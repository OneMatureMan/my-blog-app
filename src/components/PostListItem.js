import React from 'react';
import { Link } from 'react-router-dom';

const PostListItem = ({ id, title, body, createdAt }) => (
    <Link to={`edit/${id}`}>
        <p>{title} - {createdAt.valueOf()}</p>
    </Link> 
);

export default PostListItem;