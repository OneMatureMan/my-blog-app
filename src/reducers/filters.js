const filtersDefaultState = {
    text: '',
    sortBy: 'date'
}

const filtersReducer = (state = filtersDefaultState, action) => {
    switch (action.type) {
        case 'FILTER_BY_TEXT':
            return {
                ...state,
                text: action.text
            };
        case 'SORT_BY_DATE':
            return {
                ...state,
                sortBy: 'date'
            };
        case 'SORT_BY_TITLE':
            return {
                ...state,
                sortBy: 'title'
            };
        default:
            return state;
    };
};

export default filtersReducer;