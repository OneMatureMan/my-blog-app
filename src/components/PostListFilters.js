import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { filterByText, sortByDate, sortByTitle } from '../actions/filters';


export class PostListFilters extends React.Component {
    onTextChange = (e) => {
        this.props.filterByText(e.target.value);
    }

    onSortByChange = (e) => {
        if (e.target.value === 'date'){
            return this.props.sortByDate()
        }
        return this.props.sortByTitle()
    }
    render() {
        return (
            <div>
                <input 
                    className='text-input'
                    placeholder='Search posts'
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange} 
                />
                <select 
                    className='select'
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="title">Title</option>
                </select>
                <Link to='/create'>
                    <button>Add Post</button>
                </Link>
            </div>
        )
    }
}

const mapStateToProps = state => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    filterByText: (text) => dispatch(filterByText(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByTitle: () => dispatch(sortByTitle())
});



export default connect(mapStateToProps, mapDispatchToProps)(PostListFilters);