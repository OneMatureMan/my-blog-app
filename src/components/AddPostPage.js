import React from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { addPost } from '../actions/posts';


export class AddPostPage extends React.Component {

    onSubmit = (expense) => {
        this.props.addPost(expense)
        this.props.history.push('/dashboard');
    }

    render() {
        return (
            <PostForm 
                onSubmit={this.onSubmit}
            />
        );
    };
};


const mapDispatchToProps = (dispatch) => ({
    addPost : (post) => dispatch(addPost(post))
})

export default connect(undefined, mapDispatchToProps)(AddPostPage);