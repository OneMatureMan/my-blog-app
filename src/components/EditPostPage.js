import React from 'react';
import { connect } from 'react-redux';
import { editPost, removePost } from '../actions/posts';
import PostForm from './PostForm';


export class EditPostPage extends React.Component {

    onSubmit = (post) => {
        this.props.editPost(this.props.post.id,post);
        this.props.history.push('/dashboard');
    }

    onRemovePost = () => {
        this.props.removePost({ id: this.props.post.id});
        this.props.history.push('/dashboard');
    }

    render() {
        return (
            <div>
                <PostForm 
                    onSubmit={this.onSubmit}
                    post={this.props.post}
                />
                <button onClick={this.onRemovePost}>Remove Post</button>
            </div>
        );
    };
};

const mapStateToProps = (state,props) => ({
    post: state.posts.find(post => post.id === props.match.params.id)   
})


const mapDispatchToProps = (dispatch) => ({
    editPost : (id, updates) => dispatch(editPost(id, updates)),
    removePost: (id) => dispatch(removePost(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);