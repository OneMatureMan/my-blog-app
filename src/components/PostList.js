import React from 'react';
import { connect } from 'react-redux'
import PostListItem from './PostListItem';
import selectPosts from '../selectors/posts';

export const PostList = ({posts}) => (
    <div>
        <h1>THE POSTS!</h1>
        {posts.map((post) => {
            return <PostListItem key={post.id} {...post} />
        })}
    </div>
)

const mapStateToProps = state => ({
    posts: selectPosts(state.posts,state.filters)
})

export default connect(mapStateToProps)(PostList);


