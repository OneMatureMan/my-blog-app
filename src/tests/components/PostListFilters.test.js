import React from 'react';
import { shallow } from 'enzyme';
import { PostListFilters } from '../../components/PostListFilters';
import { filters, altFilters } from '../fixtures/filters';

let wrapper, filterByText, sortByDate, sortByTitle;

beforeEach(() => {
    filterByText = jest.fn();
    sortByDate = jest.fn();
    sortByTitle = jest.fn();
    wrapper = shallow(<PostListFilters 
        filters={filters} 
        filterByText={filterByText} 
        sortByDate={sortByDate} 
        sortByTitle={sortByTitle} 
    />);
})

test('Should render Post List Filters component correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('text should be empty string', () => {
    const text = wrapper.find('input').prop('value')
    expect(text).toBe('')
});

test('text should be Farsh (alt filters)', () => {
    wrapper.setProps({
        filters: altFilters
    });
    const text = wrapper.find('input').prop('value')
    expect(text).toBe('Farsh')
});

test('should change text to something different', () => {
    const value = 'different';
    wrapper.find('input').simulate('change', {
        target: { value}
    })
    expect(filterByText).toHaveBeenLastCalledWith('different');
});

test('should change sort by date to sort by title', () => {
    const value = 'title';
    wrapper.find('select').simulate('change', {
        target: { value }
    });
    expect(sortByTitle).toHaveBeenCalled();
})


test('should change sort by title to sort by date', () => {
    wrapper.setProps({
        filters: altFilters
    });
    const value = 'date';
    wrapper.find('select').simulate('change', {
        target: { value }
    });
    expect(sortByDate).toHaveBeenCalled();
});