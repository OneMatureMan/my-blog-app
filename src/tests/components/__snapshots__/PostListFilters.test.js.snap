// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Post List Filters component correctly 1`] = `
<div>
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Search expenses"
    type="text"
    value=""
  />
  <select
    className="select"
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="title"
    >
      Title
    </option>
  </select>
  <Link
    replace={false}
    to="/create"
  >
    <button>
      Add Post
    </button>
  </Link>
</div>
`;
